python/pyproject.toml

```diff
@@ -0,0 +1,145 @@
+    "Development Status :: 5 - Production/Stable",
+    "Environment :: Console",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: MIT License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 2",
+    "Programming Language :: Python :: 2.7",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.5",
+    "Programming Language :: Python :: 3.6",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Programming Language :: Python :: Implementation :: PyPy",
+    "Topic :: Software Development :: Testing",
+    "Topic :: Software Development :: Testing :: Acceptance",
+    "Topic :: Software Development :: Testing :: BDD",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "Topic :: Utilities",
+]
+
+
+[project.urls]
+homepage = "https://github.com/cucumber/tag-expressions"
+repository = "https://github.com/cucumber/tag-expressions"
+download = "https://pypi.org/project/cucumber-tag-expressions"
+documentation = "https://cucumber.io/docs/cucumber/api/#tag-expressions"
+changelog = "https://github.com/cucumber/tag-expressions/blob/main/CHANGELOG.md"
+"Issue Tracker" = "https://github.com/cucumber/tag-expressions/issues/"
+
+
+[project.optional-dependencies]
+testing = [
+    "pytest <  5.0; python_version <  '3.0'", # >= 4.2
+    "pytest >= 5.0; python_version >= '3.0'",
+    "pytest-html >= 1.19.0,<2.0; python_version <  '3.0'",
+    "pytest-html >= 2.0;         python_version >= '3.0'",
+]
+develop = [
+    "pytest <  5.0; python_version <  '3.0'", # >= 4.2
+    "pytest >= 5.0; python_version >= '3.0'",
+    "pytest-html >= 1.19.0",
+    "pytest-cov",
+    "coverage >= 5.0.3",
+    "tox >=4.20,<4.21",
+    "pylint",
+    "ruff",
+
+    # -- INVOKE SUPPORT: ----------------------------------
+    # HINT: path.py => path (python-install-package was renamed for python3)
+    "invoke >= 1.4.1",
+    "six >= 1.15.0",
+    "invoke-cleanup @ git+https://github.com/jenisys/invoke-cleanup@v0.3.7",
+    "path >= 13.1.0;    python_version >= '3.5'",
+    "path.py == 11.5.2; python_version <  '3.5'",
+
+    # -- PYTHON2 BACKPORTS:
+    "pathlib;    python_version <= '3.4'",
+    "backports.shutil_which; python_version <= '3.3'",
+    # -- CLEANUP SUPPORT: py.cleanup
+    "pycmd",
+]
+
+
+[tool.distutils.bdist_wheel]
+universal = true
+
+
+# -----------------------------------------------------------------------------
+# PACAKING TOOL SPECIFIC PARTS:
+# -----------------------------------------------------------------------------
+[tool.setuptools]
+platforms = ["any"]
+zip-safe = true
+
+# -- PREPARED:
+[tool.setuptools.dynamic]
+version = {attr = "cucumber_tag_expressions.__version__"}
```

Comment:
We are using [polyglot-release](https://github.com/cucumber/polyglot-release) to prepare a release and update the version numbers for all different languages It currently only recognizes version in the package manager. It will not update the `__version__` property.

Is there a need to also keep it as a property in the source code?

Reply:
In my opinion, yes.
It was probably overlooked in the past that the version was there.
You normally use tools like [https://pypi.org/project/bumpversion/](https://pypi.org/project/bumpversion/) or similar to accomplish this in one step.

Event if the version info is removed from Python package,
the `pyproject.toml` and/or `setup.py` should/must contain it (for readability). Note that `setup.py` will be removed in the future when Python2 compatibility of the package is dropped.

EXAMPLE:
* [repo:parse_type/.bump version.cfg](https://github.com/jenisys/parse_type/blob/main/.bumpversion.cfg)

SEE ALSO:
* [cucumber-expressions/python/pyproject.toml — version](https://github.com/cucumber/cucumber-expressions/blob/main/python/pyproject.toml#L3)

TECHNICAL ALTERNATIVE:
* [setuptools-scm](https://pypi.org/project/setuptools-scm/) uses git meta-data, like `tags`

Reply:
Thanks for explaining the different options.

It wasn't overlooked. Rather it wasn't looked at at all. Releases are made with a script. This is necessary because nobody can remember the intricacies of releasing (currently) 13 different languages.

So for the sake of the sanity of everyone involved in the release there should be exactly one conventional location for every python project in the Cucumber org to keep its version.

Currently the preferred place is directly in `pyproject.toml`.

I am open to better ways of doing this, but then I'd like to see a set of merge requests apply that change to all repositories at once.

So for now could you remove the `__version__` attribute? The other changes look like a general improvement so I think it would be okay to keep them.

Reply:
I think we should just remove the `__version__` attribute from `__init__.py`. 
The recommended way of getting the version of an installed package is to use `importlib.metadata.version(package_name)`: https://docs.python.org/3/library/importlib.metadata.html#distribution-versions

Reply:
@youtux - Can you maybe move to get this as an agreed consensus with other notable python contributors and then make a set of changes across all python implementations.

For ease (And lack of bugs), if you could make the PR's - **but not merge them** and tag one of myself or @mpkorstanje or @davidjgoss then we could liaise with yourself and others to get them all merged simultaneously whilst updating our [polyglot-release](https://github.com/cucumber/polyglot-release) repo, then this would be the smallest amount of friction.

Any help / queries, I'm in discord